### PR TITLE
Xlsbroker v4update

### PR DESCRIPTION
**Pull request by daniwelter**
v4-compatible versions of the s/sheet converter and the ingest exporter. Tested locally in dry run modes against v4 s/sheet examples - both components generate v4 schema valid json.

@danielvaughan please review

### CODE REVIEW

File: broker/stagingapi.py [hunk 1]
```diff
@@ -66,20 +66,21 @@ def deleteStagingArea(self,submissionId):
...
+        #     responseObject = json.loads(r.text)
+        #     return FileDescription(responseObject["checksums"],type,responseObject["name"],responseObject["size"],responseObject["url"], )
+        # raise ValueError('Can\'t create staging area for sub id:' +submissionId)
 
     def hasStagingArea(self, submissionId):
-        base = urlparse.urljoin( self.url, self.apiversion+'/area/'+submissionId)
-
-        r = requests.get(base, headers=self.header)
-        return r.status_code == requests.codes.ok
+        # base = urlparse.urljoin( self.url, self.apiversion+'/area/'+submissionId)
```
Comment (rolando-ebi) [inline broker/stagingapi.py]: need to uncomment this method and stageFile() above

File: flatten.py [hunk 2]
```diff
@@ -31,5 +31,5 @@ def isAtomicType(thing):
 
 def run(objectToFlatten):
     flattenedObject = collections.OrderedDict(); # maintains insertion order, which looks neater
-    flatten(unflattenedJsonObject, "", flattenedObject) # initialize the flatten() with an empty string key prefix
+    flatten(objectToFlatten, "", flattenedObject) # initialize the flatten() with an empty string key prefix
```
Comment (danielvaughan) [inline flatten.py:34]: In [ingest-archiver](https://github.com/HumanCellAtlas/ingest-archiver) we used an existing python module to flatten the JSON. Please consider using this in place of the custom flatten code specifically: flatten_json

File: broker/stagingapi.py [hunk 3]
```diff
@@ -66,20 +66,21 @@ def deleteStagingArea(self,submissionId):
...
         fileUrl = urlparse.urljoin( self.url, self.apiversion+'/area/'+submissionId+"/"+filename)
-
-        header = dict(self.header)
-        header['Content-type'] = 'application/json; dcp-type=' + type
-        r = requests.put(fileUrl,  data=json.dumps(body), headers=header)
-        if r.status_code == requests.codes.ok or requests.codes.created:
-            responseObject = json.loads(r.text)
-            return FileDescription(responseObject["checksums"],type,responseObject["name"],responseObject["size"],responseObject["url"], )
-        raise ValueError('Can\'t create staging area for sub id:' +submissionId)
+        #
```
Comment (danielvaughan) [inline broker/stagingapi.py]: Please removed commented out code if not used

**Comment by rolando-ebi** [comment]
We need to uncomment this code, I think it was temporarily commented for testing purposed

**Comment by daniwelter** [comment]
@rolando-ebi These are my fault, I will fix them right away.

File: broker/stagingapi.py (hunk 1 above)
Comment (daniwelter) [reply broker/stagingapi.py]: Done

File: broker/stagingapi.py (hunk 3 above)
Comment (daniwelter) [reply broker/stagingapi.py]: Comments removed

**Comment by daniwelter** [comment]
All accidentally left in comments should now be gone. 

Not sure what to do about the flatten.py - I didn't really touch this, apart from OCD-fixing a variable name that was throwing an error. Not sure where it's used, really.

File: broker/ingestexportservice.py [hunk 4]
```diff
@@ -113,31 +118,31 @@ def secondarySubmission(self, submissionEnvelopeUuid, analyses):
...
             analysisDssUuid = unicode(uuid.uuid4())
             analysisBundleContent = self.getBundleDocument(analysis)
             analysisFileName = "analysis_0.json" # TODO: shouldn't be hardcoded
-            fileDescription = self.writeMetadataToStaging(submissionEnvelopeUuid, analysisFileName, analysisBundleContent, "\"metadata/analysis\"")
 
             bundleManifest.fileAnalysisMap = { analysisDssUuid : [analysisUuid] }
-            filesToTransfer.append({"name":analysisFileName,
+
+            if not self.dryrun:
+                fileDescription = self.writeMetadataToStaging(submissionEnvelopeUuid, analysisFileName, analysisBundleContent, "hca-analysis")
```
Comment (rolando-ebi) [inline broker/ingestexportservice.py]: The last parameter here should be "\"metadata/analysis\""

Comment (daniwelter) [reply broker/ingestexportservice.py]: Fixed.

Comment (danielvaughan) [reply broker/ingestexportservice.py]: Ok, let's not worry about replacing flatten.py for now but flag it as something to look at next time.

Comment (danielvaughan) [reply broker/ingestexportservice.py]: Ok, let's not worry about replacing flatten.py for now but flag it as something to look at next time.

**Pull request merged**